### PR TITLE
TGP-1003: Scaffold and content for trader reference

### PR DESCRIPTION
**Pull request by vinoocs**
Ticket : https://jira.tools.tax.service.gov.uk/browse/TGP-1003
Conf : https://confluence.tools.tax.service.gov.uk/pages/viewpage.action?pageId=861962859

- Scaffold and content has been added

### CODE REVIEW

File: test/controllers/CreateRecordStartControllerSpec.scala [hunk 1]
```diff
@@ -52,7 +53,7 @@ class CreateRecordStartControllerSpec extends SpecBase {
 
         status(result) mustEqual SEE_OTHER
 
-        redirectLocation(result).value mustEqual routes.IndexController.onPageLoad.url
+        redirectLocation(result).value mustEqual routes.TraderReferenceController.onPageLoad(NormalMode).url
```
Comment (yohan-oc) [inline test/controllers/CreateRecordStartControllerSpec.scala:55]: Unit test name also needs to be updated

`"must redirect to the index controller page when the user click continue button"`

File: app/forms/TraderReferenceFormProvider.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import forms.mappings.Mappings
+import play.api.data.Form
+
+class TraderReferenceFormProvider @Inject() extends Mappings {
+
+  def apply(): Form[String] =
+    Form(
+      "value" -> text("traderReference.error.required")
+        .verifying(maxLength(100, "traderReference.error.length"))
```
Comment (yohan-oc) [inline app/forms/TraderReferenceFormProvider.scala]: 512 characters?

File: test/controllers/CreateRecordStartControllerSpec.scala (hunk 1 above)
Comment (vinoocs) [reply test/controllers/CreateRecordStartControllerSpec.scala:55]: Updated :D

File: app/forms/TraderReferenceFormProvider.scala (hunk 2 above)
Comment (vinoocs) [reply app/forms/TraderReferenceFormProvider.scala]: Updated!

File: app/navigation/Navigator.scala [hunk 3]
```diff
@@ -36,6 +36,8 @@ class Navigator @Inject() () {
     case HasGoodsDescriptionPage => navigateFromHasGoodsDescription
     case GoodsDescriptionPage    => _ => routes.CountryOfOriginController.onPageLoad(NormalMode)
     case CommodityCodePage       => _ => routes.HasCorrectGoodsController.onPageLoad(NormalMode)
+    case CountryOfOriginPage     => _ => routes.CommodityCodeController.onPageLoad(NormalMode)
```
Comment (besscerule) [inline app/navigation/Navigator.scala:39]: Can we have these rows in the order of the pages (so one lead on from the other so it looks nicer)

**Review by Ewan-James-Donovan — DISMISSED** [review]

File: app/viewmodels/checkAnswers/TraderReferenceSummary.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+object TraderReferenceSummary {
+
+  def row(answers: UserAnswers)(implicit messages: Messages): Option[SummaryListRow] =
+    answers.get(TraderReferencePage).map { answer =>
+      SummaryListRowViewModel(
+        key = "traderReference.checkYourAnswersLabel",
+        value = ValueViewModel(HtmlFormat.escape(answer).toString),
+        actions = Seq(
+          ActionItemViewModel("site.change", routes.TraderReferenceController.onPageLoad(CheckMode).url)
+            .withVisuallyHiddenText(messages("traderReference.change.hidden"))
```
Comment (besscerule) [inline app/viewmodels/checkAnswers/TraderReferenceSummary.scala:37]: you need messages for this

File: app/viewmodels/checkAnswers/TraderReferenceSummary.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import uk.gov.hmrc.govukfrontend.views.viewmodels.summarylist.SummaryListRow
+import viewmodels.govuk.summarylist._
+import viewmodels.implicits._
+
+object TraderReferenceSummary {
+
+  def row(answers: UserAnswers)(implicit messages: Messages): Option[SummaryListRow] =
+    answers.get(TraderReferencePage).map { answer =>
+      SummaryListRowViewModel(
+        key = "traderReference.checkYourAnswersLabel",
```
Comment (besscerule) [inline app/viewmodels/checkAnswers/TraderReferenceSummary.scala:33]: you need messages for this

**Review by besscerule — CHANGES_REQUESTED** [review]

**Review by yohan-oc — DISMISSED** [review]
LGTM

File: app/navigation/Navigator.scala (hunk 3 above)
Comment (vinoocs) [reply app/navigation/Navigator.scala:39]: Done it :D

File: app/viewmodels/checkAnswers/TraderReferenceSummary.scala (hunk 4 above)
Comment (vinoocs) [reply app/viewmodels/checkAnswers/TraderReferenceSummary.scala:37]: It's a generated view model + hidden. Do we update it

File: app/viewmodels/checkAnswers/TraderReferenceSummary.scala (hunk 5 above)
Comment (vinoocs) [reply app/viewmodels/checkAnswers/TraderReferenceSummary.scala:33]: same as above.

File: app/viewmodels/checkAnswers/TraderReferenceSummary.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import pages.TraderReferencePage
+import play.api.i18n.Messages
+import play.twirl.api.HtmlFormat
+import uk.gov.hmrc.govukfrontend.views.viewmodels.summarylist.SummaryListRow
+import viewmodels.govuk.summarylist._
+import viewmodels.implicits._
+
+object TraderReferenceSummary {
+
+  def row(answers: UserAnswers)(implicit messages: Messages): Option[SummaryListRow] =
```
Comment (liandra-giacomine) [inline app/viewmodels/checkAnswers/TraderReferenceSummary.scala:30]: Confused - what is TraderReferenceSummary for?

File: app/viewmodels/checkAnswers/TraderReferenceSummary.scala (hunk 4 above)
Comment (besscerule) [reply app/viewmodels/checkAnswers/TraderReferenceSummary.scala:37]: Yes, see the other examples in messages.en

**Review by besscerule — APPROVED** [review]

File: app/viewmodels/checkAnswers/TraderReferenceSummary.scala (hunk 6 above)
Comment (vinoocs) [reply app/viewmodels/checkAnswers/TraderReferenceSummary.scala:30]: @liandra-giacomine : All these auto generated summary objects are used in cya page-load to display the data without much effort. This is generated for all pages with check modes. Pasting the code example from other cya controller

 def onPageLoad(): Action[AnyContent] = (identify andThen getData andThen requireData) { implicit request =>
    TraderProfile.build(request.userAnswers, request.eori) match {
      case Right(_)     =>
        val list = SummaryListViewModel(
          rows = Seq(
            UkimsNumberSummary.row(request.userAnswers),
            HasNirmsSummary.row(request.userAnswers),
            NirmsNumberSummary.row(request.userAnswers),
            HasNiphlSummary.row(request.userAnswers),

**Review by liandra-giacomine — APPROVED** [review]

**Pull request merged**